src/gwt/panmirror/src/editor/src/optional/ace/render_queue.ts

```diff
@@ -0,0 +1,215 @@
+/*
+ * render_queue.ts
+ *
+ * Copyright (C) 2020 by RStudio, PBC
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+import { CodeBlockNodeView } from './ace';
+
+/**
+ * Represents a queue of Ace editor instances that are rendered asynchronously.
+ */
+export class AceRenderQueue {
+  private readonly renderQueue: CodeBlockNodeView[] = [];
+
+  private renderCompleted: boolean = false;
+  private renderTimer: number = 0;
+  private container?: HTMLElement;
+  private needsSort: boolean = true;
+  private bypass: number = 5;
+  private observer?: IntersectionObserver;
+  private visible: boolean = true;
+
+  /**
+   * Sets (or replaces) the scroll container hosting the render queue. The
+   * scroll container is used to prioritize the queue (i.e. objects in the
+   * viewport or close to it are to be given more priority).
+   * 
+   * @param container The HTML element of the scroll container
+   */
+  public setContainer(container: HTMLElement) {
+
+    // Handler for scroll events in the container
+    const handleScroll = (evt: Event) => {
+      // Whenever a scroll event occurs, we need to re-sort the queue so that
+      // objects in or closest to the viewport are given priority.
+      this.needsSort = true;
+
+      // If we don't think we're visible but we're scrolling and have height,
+      // then we are in fact visible. (This catches a case where the
+      // intsersection observer created below doesn't fire for visiblity
+      // changes.)
+      if (!this.visible && this.container && this.container.offsetHeight > 0) {
```

Comment:
IIUC, querying `offsetHeight` can trigger reflow, so this might be expensive to do in a scroll handler. Not sure if this is an issue in practice for us but wonder if this is worth looking at?

Reply:
That's a good question. In practice it's not an issue. `this.visible` is `true` almost all the time (short circuiting the rest of the logic), and when it's not, the container doesn't get scroll events -- so the only time we'd ever check the `offsetHeight` is when we are in an unexpected case (we are getting scroll events somehow even though we don't think we are visible). In this case we confirm that we are actually visible by checking the offsetHeight, after which `visible` becomes true so we don't check again.